R/motifs.R

```diff
@@ -125,10 +126,12 @@ dyad.census <- function(graph) { # nocov start
 #' motifs(g, 3)
 #' count_motifs(g, 3)
 #' sample_motifs(g, 3)
-motifs <- function(graph, size = 3, cut.prob = rep(0, size)) {
+motifs <- function(graph, size = 3, cut.prob = NULL) {
   ensure_igraph(graph)
-  cut.prob <- as.numeric(cut.prob)
-  if (length(cut.prob) != size) {
+
+  if (!is.null(cut.prob)) cut.prob <- as.numeric(cut.prob)
+
+  if (!is.null(cut.prob) && length(cut.prob) != size) {
```

Comment:
Please help me understand the code below. Is it repeating the last entry to fill up the vector until it reaches `size`?

What if it is longer than `size`? That should be an error (and I think the C core will take care of that, but I'd need to check).

Reply:
This just errors:

```r
motifs(make_ring(10), cut.prob=0.1)
Error in motifs(make_ring(10), cut.prob = 0.1) : 
  At vendor/cigraph/src/misc/motifs.c:156 : Cut probability vector size (0) must agree with motif size (3). Invalid value
```

Why does it not trigger the `length(cut.prob) != size` code path?

I think I'm just too tired for this right now :(

Reply:
This code path _is_ triggered, but it produces an incorrect vector.

I know this is not your code, but can you fix it @maelle ?

My recommendation is:

 1. Either go with what this code seems to intend, i.e. repeat the last element as many times as necessary to reach `size`.
 2. Or keep things simple and not repeat anything _except_ a scalar. A scalar should be treated as a vector of identical elements of length `size`.

No preference as to which from my size (maybe 1 is a tiny bit nicer, but also overly complicated?)

Whatever you choose, can you document it? Also, can you make sure it's implemented in _all_ motif functions, not just this one?

Reply:
So do you mean

- error if `cut.prob` is not of length either 1 or the same as size?
- if it is of length 1, repeat it size times?

Reply:
That is option 2.

Let me give examples for option 1 and option 2, assuming `size == 4`.

Option 1:

Input: `0.3`, interpreted as `0.3 0.3 0.3 0.3`

Input: `0.3 0.4`, interpreted as `0.3 0.4 0.4 0.4`

Input: `0.1 0.2 0.3 0.4 0.5` (longer than size 4), should trigger an error, which can be achieved by passing to the C core as-is

So, repeat the last entry to fill to size.

Option 2:

Input: `0.3`, interpreted as `0.3 0.3 0.3 0.3`

Input: `0.3 0.4`, passed to the C core as-is, which will trigger an error.

Reply:
@krlmlr any opinion here?